webservices/server-integration/src/main/java/org/jboss/as/webservices/security/ElytronSecurityDomainContextImpl.java

```diff
@@ -91,8 +93,18 @@ public void runAs(Callable<Void> action) throws Exception {
         }
     }
     @Override
-    public void pushSubjectContext(Subject arg0, Principal arg1, Object arg2) {
-
+    public void pushSubjectContext(Subject subject, Principal pincipal, Object credential) {
+        AccessController.doPrivileged(new PrivilegedAction<Void>() {
+            public Void run() {
+                if (credential != null) {
+                    subject.getPrivateCredentials().add(credential);
+                }
+                SecurityIdentity securityIdentity = SubjectUtil.convertToSecurityIdentity(subject, pincipal, securityDomain,
+                        "ejb");
+                currentIdentity.set(securityIdentity);
```

Comment:
Then https://github.com/wildfly/wildfly/blob/85d121b86043c75d145635641a5549f5eda0c13d/webservices/server-integration/src/main/java/org/jboss/as/webservices/security/ElytronSecurityDomainContextImpl.java#L80 is redundant, right?

Reply:
This will set the converted thread local securityIdentity first , then runAs with this securityIdentiy for EJB web method invocation.

Reply:
You now have two places in which the currentIdentity is set, which makes no sense.

Reply:
That still doesn't explain https://github.com/wildfly/wildfly/blob/4d4c7705d955b8ff7f2f9e7af50bf6784a944b0d/webservices/server-integration/src/main/java/org/jboss/as/webservices/security/ElytronSecurityDomainContextImpl.java#L80 vs https://github.com/wildfly/wildfly/blob/4d4c7705d955b8ff7f2f9e7af50bf6784a944b0d/webservices/server-integration/src/main/java/org/jboss/as/webservices/security/ElytronSecurityDomainContextImpl.java#L104

Reply:
Reading the code my interpretation is either isValid() OR pushSubjectContext() can be called - runAs() is subsequently called which will clear any previously associated identity on the ThreadLocal.

But looking at this we do need to be 100% that nothing would prevent runAs() being called after isValid() or pushSubjectContext() has returned.

Additionally if we are relying on the runAd method popping the identity ci.runAs probably needs wrapping in a try / finally.

Reply:
> Reading the code my interpretation is either isValid() OR pushSubjectContext() can be called - runAs() is subsequently called which will clear any previously associated identity on the ThreadLocal.
But looking at this we do need to be 100% that nothing would prevent runAs() being called after isValid() or pushSubjectContext() has returned.

Now these lines are called in CXF's interceptors which has the phase concept like our deployment phase , this can make sure runAs() is called always after pushSubjectContext().

> Additionally if we are relying on the runAd method popping the identity ci.runAs probably needs wrapping in a try / finally.

 This is better to execute in a try/finally block. Fixed.

Reply:
> Now these lines are called in CXF's interceptors which has the phase concept like our deployment phase , this can make sure runAs() is called always after pushSubjectContext().
But then what is the use of `currentIdentity.set(identity); ` in `isValid`?

Reply:
I agree with Carlo, the set(identity) in isValid should be removed; pushSubjectContext is what is supposed to propagate the identity.

Reply:
> But looking at this we do need to be 100% that nothing would prevent runAs() being called after isValid() or pushSubjectContext() has returned.

This is a valid concern in case something weird happen within JBossWS and the request processing fails in an unexpected way. Honestly I think I'm seeing the same issue with the old security implementation too...
I've been discussing this with Jim this morning, we should likely have a safety net implemented in the entry point for each request in jbossws, which would end up calling the SecurityDomainContext here to cleanup the threadlocal association. Not sure if this should go into a separate jira & PR.